Split string literals which are longer than 2^16 characters

I previously had an insert (on db creation) adding 505 rows of data. One column had dummy data (an empty string) and I changed it add ~100 characters per row. After that building the APK started failing with the following error:

```
...
    ACONST_NULL
    ICONST_1
    ACONST_NULL
   L219
    LINENUMBER 1808 L219
    INVOKESTATIC kotlin/text/StringsKt.trimMargin$default (Ljava/lang/String;Ljava/lang/String;ILjava/lang/Object;)Ljava/lang/String;
    ICONST_0
    ACONST_NULL
    BIPUSH 8
    ACONST_NULL
   L220
    LINENUMBER 1799 L220
    INVOKESTATIC com/squareup/sqldelight/db/SqlDriver$DefaultImpls.execute$default (Lcom/squareup/sqldelight/db/SqlDriver;Ljava/lang/Integer;Ljava/lang/String;ILkotlin/jvm/functions/Function1;ILjava/lang/Object;)V
   L221
    LINENUMBER 1809 L221
    RETURN
   L222
    LOCALVARIABLE this Lghplayer2/android/north101/gloomplayer2/Database$Schema; L0 L222 0
    LOCALVARIABLE driver Lcom/squareup/sqldelight/db/SqlDriver; L0 L222 1
    MAXSTACK = 7
    MAXLOCALS = 2

Cause: java.lang.IllegalArgumentException
Element is unknownThe root cause was thrown at: ByteVector.java:213
	at org.jetbrains.kotlin.codegen.TransformationMethodVisitor.visitEnd(TransformationMethodVisitor.java:88)
	at org.jetbrains.org.objectweb.asm.MethodVisitor.visitEnd(MethodVisitor.java:878)
	at org.jetbrains.kotlin.codegen.FunctionCodegen.endVisit(FunctionCodegen.java:992)
	... 62 more
Caused by: java.lang.IllegalArgumentException
	at org.jetbrains.org.objectweb.asm.ByteVector.putUTF8(ByteVector.java:213)
	at org.jetbrains.org.objectweb.asm.ClassWriter.newUTF8(ClassWriter.java:1141)
	at org.jetbrains.org.objectweb.asm.ClassWriter.newStringishItem(ClassWriter.java:1163)
	at org.jetbrains.org.objectweb.asm.ClassWriter.newConstItem(ClassWriter.java:1091)
	at org.jetbrains.org.objectweb.asm.MethodWriter.visitLdcInsn(MethodWriter.java:1165)
	at org.jetbrains.org.objectweb.asm.MethodVisitor.visitLdcInsn(MethodVisitor.java:595)
	at org.jetbrains.org.objectweb.asm.tree.LdcInsnNode.accept(LdcInsnNode.java:71)
	at org.jetbrains.org.objectweb.asm.tree.InsnList.accept(InsnList.java:162)
	at org.jetbrains.org.objectweb.asm.tree.MethodNode.accept(MethodNode.java:817)
	at org.jetbrains.kotlin.codegen.TransformationMethodVisitor.visitEnd(TransformationMethodVisitor.java:68)
	... 64 more
```

I was able to fix it by breaking it up into multiple inserts.

--- a/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/compiler/KotlinPoet.kt
+++ b/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/compiler/KotlinPoet.kt
@@ -1,0 +1,31 @@
+package com.squareup.sqldelight.core.compiler
+
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.MemberName
+import com.squareup.kotlinpoet.buildCodeBlock
+
+/**
+ * The maximum string length that we can emit as a literal. This value is 2^15 instead of the real
+ * value, 2^16, to account for the indentation that KotlinPoet will add by using `trimMargin`.
+ */
+private const val MAX_STRING_LENGTH = 32_768
+
+private val BUILD_STRING = MemberName("kotlin.text", "buildString")
+
+/**
+ * Returns a [CodeBlock] which contains a representation of [this] as a literal. If longer than
+ * the maximum allowed length for a string, it will be broken up into multiple literals with code
+ * to re-assemble at runtime.
+ */
+internal fun String.toCodeLiteral(): CodeBlock {
+  if (length < MAX_STRING_LENGTH) {
+    return CodeBlock.of("%S", this)
+  }
+  return buildCodeBlock {
+    add("%M(%L) {\n", BUILD_STRING, length)
+    for (i in 0 until length step MAX_STRING_LENGTH) {
+      add("append(%S)\n", substring(i, (i + MAX_STRING_LENGTH).coerceAtMost(length)))
+    }
+    add("}")
+  }
+}

--- a/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/compiler/QueryWrapperGenerator.kt
+++ b/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/compiler/QueryWrapperGenerator.kt
@@ -174,7 +174,7 @@ internal class QueryWrapperGenerator(
 
     sourceFolders.flatMap { it.findChildrenOfType<SqlDelightFile>() }
         .forInitializationStatements { sqlText ->
-          createFunction.addStatement("$DRIVER_NAME.execute(null, %S, 0)", sqlText)
+          createFunction.addStatement("$DRIVER_NAME.execute(null, %L, 0)", sqlText.toCodeLiteral())
         }
 
     var maxVersion = 1

--- a/sqldelight-compiler/src/test/kotlin/com/squareup/sqldelight/core/QueryWrapperTest.kt
+++ b/sqldelight-compiler/src/test/kotlin/com/squareup/sqldelight/core/QueryWrapperTest.kt
@@ -365,4 +365,98 @@ class QueryWrapperTest {
         |
         """.trimMargin())
   }
+
+  @Test fun `string longer than 2^16 is chunked`() {
+    val sqString = buildString {
+      append("""
+        |CREATE TABLE class_ability_test (
+        |  id TEXT PRIMARY KEY NOT NULL,
+        |  class_id TEXT NOT NULL,
+        |  name TEXT NOT NULL,
+        |  level_id INTEGER NOT NULL DEFAULT 1,
+        |  special TEXT,
+        |  url TEXT NOT NULL
+        |);
+        |
+        |INSERT INTO class_ability_test(id, class_id, name, level_id, special, url)
+        |VALUES
+        """.trimMargin())
+      repeat(500) {
+        if (it > 0) append(',')
+        append("\n  ('class_01_ability_$it', 'class_01', 'aaaaaaaaaaaaaaa', 1, NULL, 'https://stuff.example.com/this/is/a/bunch/of/path/data/class_01_ability_$it.png')")
+      }
+      append(';')
+    }
+    val result = FixtureCompiler.compileSql(sqString, tempFolder)
+
+    assertThat(result.errors).isEmpty()
+
+    val queryWrapperFile = result.compilerOutput[File(result.outputDirectory, "com/example/testmodule/TestDatabaseImpl.kt")]
+    assertThat(queryWrapperFile.toString()).apply {
+      startsWith("""
+        |package com.example.testmodule
+        |
+        |import com.example.TestDatabase
+        |import com.example.TestQueries
+        |import com.squareup.sqldelight.TransacterImpl
+        |import com.squareup.sqldelight.db.SqlDriver
+        |import kotlin.Int
+        |import kotlin.reflect.KClass
+        |import kotlin.text.buildString
+        |
+        |internal val KClass<TestDatabase>.schema: SqlDriver.Schema
+        |    get() = TestDatabaseImpl.Schema
+        |
+        |internal fun KClass<TestDatabase>.newInstance(driver: SqlDriver): TestDatabase =
+        |        TestDatabaseImpl(driver)
+        |
+        |private class TestDatabaseImpl(driver: SqlDriver) : TransacterImpl(driver), TestDatabase {
+        |    override val testQueries: TestQueriesImpl = TestQueriesImpl(this, driver)
+        |
+        |    object Schema : SqlDriver.Schema {
+        |        override val version: Int
+        |            get() = 1
+        |
+        |        override fun create(driver: SqlDriver) {
+        |            driver.execute(null, ""${'"'}
+        |                    |CREATE TABLE class_ability_test (
+        |                    |  id TEXT PRIMARY KEY NOT NULL,
+        |                    |  class_id TEXT NOT NULL,
+        |                    |  name TEXT NOT NULL,
+        |                    |  level_id INTEGER NOT NULL DEFAULT 1,
+        |                    |  special TEXT,
+        |                    |  url TEXT NOT NULL
+        |                    |)
+        |                    ""${'"'}.trimMargin(), 0)
+        |            driver.execute(null, buildString(75360) {
+        |                    append(""${'"'}
+        |                    |INSERT INTO class_ability_test(id, class_id, name, level_id, special, url)
+        |                    |VALUES
+        |                    |  ('class_01_ability_0', 'class_01', 'aaaaaaaaaaaaaaa', 1, NULL, 'https://stuff.example.com/this/is/a/bunch/of/path/data/class_01_ability_0.png')
+        """.trimMargin())
+      contains("""
+        |                    ""${'"'}.trimMargin())
+        |                    append(""${'"'}
+      """.trimMargin())
+      endsWith("""
+        |                    |  ('class_01_ability_499', 'class_01', 'aaaaaaaaaaaaaaa', 1, NULL, 'https://stuff.example.com/this/is/a/bunch/of/path/data/class_01_ability_499.png')
+        |                    ""${'"'}.trimMargin())
+        |                    }, 0)
+        |        }
+        |
+        |        override fun migrate(
+        |            driver: SqlDriver,
+        |            oldVersion: Int,
+        |            newVersion: Int
+        |        ) {
+        |        }
+        |    }
+        |}
+        |
+        |private class TestQueriesImpl(private val database: TestDatabaseImpl, private val driver: SqlDriver)
+        |        : TransacterImpl(driver), TestQueries
+        |
+        """.trimMargin())
+    }
+  }
 }

